File: src/gear/pkgbuild_analyzer.py
import os
import re
import subprocess
import logging
import difflib # For diffing PKGBUILDs
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass, field

# Basic logging configuration for this module
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger("pkgbuild_analyzer")

@dataclass
class PkgbuildFunction:
    """Represents a shell function within a PKGBUILD."""
    name: str
    content: str
    start_line: int
    end_line: int

@dataclass
class PkgbuildMetadata:
    """Core metadata extracted from a PKGBUILD."""
    pkgname: str
    pkgver: str
    pkgrel: str = "1"
    epoch: Optional[str] = None
    arch: List[str] = field(default_factory=list)
    url: Optional[str] = None
    license: Optional[str] = None
    depends: List[str] = field(default_factory=list)
    makedepends: List[str] = field(default_factory=list)
    checkdepends: List[str] = field(default_factory=list)
    optdepends: List[str] = field(default_factory=list)
    source: List[str] = field(default_factory=list) # Raw source array entries
    functions: Dict[str, PkgbuildFunction] = field(default_factory=dict)

@dataclass
class PkgbuildComparisonResult:
    """Result of comparing two PKGBUILDs."""
    diff: str # Unified diff string
    has_changes: bool
    critical_sections_changed: List[str] # e.g., ['source', 'prepare']
    version_diff: Tuple[Optional[str], Optional[str]] = (None, None) # (local_ver, upstream_ver)

@dataclass
class DependencyAnalysisResult:
    """Result of analyzing package dependencies."""
    package_name: str
    current_pkgver: str
    dependencies: Dict[str, List[str]] = field(default_factory=dict) # e.g., {'depends': [...], 'makedepends': [...]}
    missing_deps: List[str] = field(default_factory=list)
    installed_deps: List[str] = field(default_factory=list)
    optional_deps_available: List[str] = field(default_factory=list)


class PKGBUILDAnalyzer:
    """
    Provides specialized analysis capabilities for PKGBUILD files,
    including detailed parsing, diffing against upstream versions,
    and dependency analysis.
    """

    # Regex for extracting array-like variables (e.g., depends, source)
    # This regex is simplified and might need refinement for edge cases.
    _array_var_re = re.compile(r'^\s*(?P<var_name>\w+)\s*=\s*\((?P<content>[^)]*)\)', re.MULTILINE | re.DOTALL)
    _single_var_re = re.compile(r'^\s*(?P<var_name>\w+)\s*=\s*(?P<content>.*?)\s*$', re.MULTILINE)
    _function_re = re.compile(r'^\s*(?P<func_name>\w+)\s*\(\s*\)\s*{', re.MULTILINE)
    _func_end_re = re.compile(r'^\s*}', re.MULTILINE)

    def __init__(self, upstream_checker: Optional[Any] = None):
        """
        Initializes the PKGBUILDAnalyzer.

        Args:
            upstream_checker: An instance of UpstreamChecker to fetch upstream PKGBUILDs.
        """
        self.upstream_checker = upstream_checker
        logger.info("PKGBUILDAnalyzer initialized.")

    def parse_pkgbuild_detailed(self, pkgbuild_path: str) -> Optional[PkgbuildMetadata]:
        """
        [ ] Implement focused extraction of critical parts (`source=`, `prepare()`, `build()`, etc.).
        Parses a PKGBUILD file to extract all metadata variables and function contents.

        Args:
            pkgbuild_path: The path to the PKGBUILD file.

        Returns:
            A PkgbuildMetadata object if parsing is successful, None otherwise.
        """
        if not os.path.exists(pkgbuild_path):
            logger.error(f"PKGBUILD not found: {pkgbuild_path}")
            return None

        try:
            with open(pkgbuild_path, 'r', encoding='utf-8') as f:
                content_lines = f.readlines()
            full_content = "".join(content_lines)

            # Remove comments for easier parsing of variables/functions
            content_no_comments = re.sub(r'#.*$', '', full_content, flags=re.MULTILINE)

            metadata = PkgbuildMetadata(
                pkgname="unknown", pkgver="unknown", arch=[],
                depends=[], makedepends=[], checkdepends=[], optdepends=[], source=[]
            )

            # Extract basic variables (pkgname, pkgver, pkgrel, epoch, url, license)
            for var_name in ['pkgname', 'pkgver', 'pkgrel', 'epoch', 'url', 'license', 'arch']:
                match = re.search(r'^\s*' + var_name + r'\s*=\s*(?P<value>.*?)\s*$', content_no_comments, re.MULTILINE)
                if match:
                    value = match.group('value').strip().strip('\'"()') # Strip quotes and parentheses
                    if var_name == 'pkgname': metadata.pkgname = value
                    elif var_name == 'pkgver': metadata.pkgver = value
                    elif var_name == 'pkgrel': metadata.pkgrel = value
                    elif var_name == 'epoch': metadata.epoch = value
                    elif var_name == 'url': metadata.url = value
                    elif var_name == 'license': metadata.license = value
                    elif var_name == 'arch': metadata.arch = [a.strip() for a in value.split()]

            # Extract array variables (depends, makedepends, checkdepends, optdepends, source)
            # This logic needs to handle both single-line and multi-line arrays.
            for var_match in self._array_var_re.finditer(content_no_comments):
                var_name = var_match.group('var_name')
                raw_content = var_match.group('content')
                items = [item.strip().strip('\'"') for item in raw_content.split() if item.strip()]

                if var_name == 'depends': metadata.depends.extend(items)
                elif var_name == 'makedepends': metadata.makedepends.extend(items)
                elif var_name == 'checkdepends': metadata.checkdepends.extend(items)
                elif var_name == 'optdepends': metadata.optdepends.extend(items)
                elif var_name == 'source': metadata.source.extend(items)

            # Re-process source to substitute variables using extracted metadata
            resolved_sources = []
            for src_entry in metadata.source:
                resolved_src = src_entry.replace('$pkgname', metadata.pkgname).replace('$pkgver', metadata.pkgver)
                # Apply further variable substitutions if necessary (e.g., $arch, custom variables)
                resolved_sources.append(resolved_src)
            metadata.source = resolved_sources


            # Extract shell functions (prepare, build, package, check, pkgver)
            in_function = False
            current_func_name = ""
            func_content_lines = []
            func_start_line = -1

            for i, line in enumerate(content_lines):
                func_start_match = self._function_re.match(line)
                func_end_match = self._func_end_re.match(line)

                if func_start_match:
                    if in_function: # Handle case where a function might not have a closing brace
                        logger.warning(f"PKGBUILD parsing: Function '{current_func_name}' did not have closing brace before new function at line {i+1}.")
                        metadata.functions[current_func_name] = PkgbuildFunction(
                            name=current_func_name, content="".join(func_content_lines),
                            start_line=func_start_line, end_line=i-1
                        )

                    in_function = True
                    current_func_name = func_start_match.group('func_name')
                    func_content_lines = []
                    func_start_line = i + 1 # Content starts on next line (after function header)

                elif in_function and func_end_match:
                    metadata.functions[current_func_name] = PkgbuildFunction(
                        name=current_func_name, content="".join(func_content_lines),
                        start_line=func_start_line, end_line=i + 1 # Include the closing brace line
                    )
                    in_function = False
                    current_func_name = ""
                    func_content_lines = []
                    func_start_line = -1

                elif in_function:
                    func_content_lines.append(line)

            # Handle case where the last function might not have a closing brace
            if in_function and current_func_name:
                metadata.functions[current_func_name] = PkgbuildFunction(
                    name=current_func_name, content="".join(func_content_lines),
                    start_line=func_start_line, end_line=len(content_lines)
                )


            return metadata

        except Exception as e:
            logger.exception(f"Error detailed parsing PKGBUILD at {pkgbuild_path}: {e}")
            return None


    def compare_pkgbuilds(self, local_pkgbuild_path: str, upstream_pkgbuild_path: str) -> Optional[PkgbuildComparisonResult]:
        """
        [ ] Develop intelligent diff system to review changes between local and upstream PKGBUILD.
        Compares a local PKGBUILD with an upstream version, generating a detailed diff
        and identifying changes in critical sections.

        Args:
            local_pkgbuild_path: Path to the local PKGBUILD.
            upstream_pkgbuild_path: Path to the upstream PKGBUILD.

        Returns:
            A PkgbuildComparisonResult object, or None if files cannot be read.
        """
        if not os.path.exists(local_pkgbuild_path):
            logger.error(f"Local PKGBUILD not found: {local_pkgbuild_path}")
            return None
        if not os.path.exists(upstream_pkgbuild_path):
            logger.error(f"Upstream PKGBUILD not found: {upstream_pkgbuild_path}")
            return None

        try:
            with open(local_pkgbuild_path, 'r', encoding='utf-8') as f:
                local_lines = f.readlines()
            with open(upstream_pkgbuild_path, 'r', encoding='utf-8') as f:
                upstream_lines = f.readlines()

            # Generate unified diff
            diff_generator = difflib.unified_diff(
                local_lines, upstream_lines,
                fromfile=f"a/{os.path.basename(local_pkgbuild_path)}",
                tofile=f"b/{os.path.basename(upstream_pkgbuild_path)}",
                lineterm=''
            )
            diff_content = "".join(list(diff_generator))
            has_changes = bool(diff_content)

            # Identify changes in critical sections (conceptual for now)
            # This would require detailed parsing of both PKGBUILDs and then comparing
            # the content of specific functions/variables like 'source', 'prepare()', 'package()'.
            critical_sections_changed = []
            if has_changes:
                # Placeholder logic: if diff exists, assume some critical sections *might* have changed
                # A real implementation would parse and compare 'source', 'url', and function contents.
                if "source=" in diff_content or "prepare()" in diff_content or "package()" in diff_content:
                    critical_sections_changed.append("source")
                    critical_sections_changed.append("prepare_function")
                    critical_sections_changed.append("package_function") # Simplified detection

            # Get version diff
            local_meta = self.parse_pkgbuild_detailed(local_pkgbuild_path)
            upstream_meta = self.parse_pkgbuild_detailed(upstream_pkgbuild_path)

            local_ver = f"{local_meta.pkgver}-{local_meta.pkgrel}" if local_meta else "N/A"
            upstream_ver = f"{upstream_meta.pkgver}-{upstream_meta.pkgrel}" if upstream_meta else "N/A"


            return PkgbuildComparisonResult(
                diff=diff_content,
                has_changes=has_changes,
                critical_sections_changed=critical_sections_changed,
                version_diff=(local_ver, upstream_ver)
            )

        except Exception as e:
            logger.exception(f"Error comparing PKGBUILDs: {e}")
            return None


    def analyze_dependencies(self, pkgbuild_path: str) -> Optional[DependencyAnalysisResult]:
        """
        [ ] Create analysis of dependencies and relations between packages.
        Analyzes the dependencies of a PKGBUILD, comparing them against installed packages.
        This is a blocking call and should be executed in a separate thread/process.

        Args:
            pkgbuild_path: Path to the PKGBUILD file.

        Returns:
            A DependencyAnalysisResult object, or None if analysis fails.
        """
        pkg_metadata = self.parse_pkgbuild_detailed(pkgbuild_path)
        if not pkg_metadata:
            logger.error(f"Could not parse PKGBUILD for dependency analysis: {pkgbuild_path}")
            return None

        logger.info(f"Analyzing dependencies for {pkg_metadata.pkgname} (blocking)...")
        result = DependencyAnalysisResult(pkg_metadata.pkgname, pkg_metadata.pkgver)
        result.dependencies = {
            'depends': pkg_metadata.depends,
            'makedepends': pkg_metadata.makedepends,
            'checkdepends': pkg_metadata.checkdepends,
            'optdepends': pkg_metadata.optdepends,
        }

        all_required_deps = set(pkg_metadata.depends + pkg_metadata.makedepends + pkg_metadata.checkdepends)

        # Fetch installed packages (this is a blocking operation)
        installed_packages_output = subprocess.run(
            ['pacman', '-Q', 'q'], # List installed packages in quiet mode
            capture_output=True, text=True, check=False, timeout=10
        )
        installed_packages = set(installed_packages_output.stdout.splitlines())

        # Check for missing dependencies
        for dep in all_required_deps:
            # Simple check: does the dependency name exist in installed packages?
            # More complex: handle version constraints (e.g., 'package>=1.0'), virtual packages
            if dep not in installed_packages:
                result.missing_deps.append(dep)
            else:
                result.installed_deps.append(dep)

        # Check for available optional dependencies
        for optdep in pkg_metadata.optdepends:
            if optdep in installed_packages: # Check if the optional dep is installed
                result.optional_deps_available.append(optdep)

        return result

    def verify_version_compatibility(self, pkgbuild_path: str) -> Tuple[bool, str]:
        """
        [ ] Implement version compatibility and system requirements verification.
        Verifies if the PKGBUILD's specified version and architecture are compatible
        with the current system. (Conceptual for now).

        Args:
            pkgbuild_path: Path to the PKGBUILD file.

        Returns:
            A tuple (is_compatible: bool, message: str).
        """
        metadata = self.parse_pkgbuild_detailed(pkgbuild_path)
        if not metadata:
            return False, "Could not parse PKGBUILD metadata."

        current_arch = subprocess.run(['uname', '-m'], capture_output=True, text=True, check=True).stdout.strip()

        # Check architecture compatibility
        if "any" not in metadata.arch and current_arch not in metadata.arch:
            return False, f"Architecture '{current_arch}' not supported by PKGBUILD (supports: {', '.join(metadata.arch)})."

        # TODO: Implement more advanced version compatibility checks
        # - Check minimum kernel version (if specified in PKGBUILD or implicit for dependencies)
        # - Check library versions (e.g., 'glibc>=2.30') by parsing dependencies and system status.

        return True, "Version and architecture appear compatible."

    def integrate_with_security_analysis(self, pkgbuild_path: str, security_analyzer: Any) -> Any:
        """
        [ ] Develop integration with the security system for PKGBUILD-specific risk analysis.
        Delegates to the SecurityAnalyzer for in-depth risk analysis of the PKGBUILD.
        """
        if not security_analyzer:
            logger.error("SecurityAnalyzer instance not provided.")
            return None

        logger.info(f"Integrating PKGBUILD-specific risk analysis for {pkgbuild_path}")
        return security_analyzer.analyze_pkgbuild(pkgbuild_path)


# Example Usage (for testing this module directly)
if __name__ == "__main__":
    logging.getLogger().setLevel(logging.DEBUG)

    # Create dummy PKGBUILD files for testing
    pkgbuild_content_local = """
    # Contributor: Local User
    pkgname=my-app
    pkgver=1.0.0
    pkgrel=1
    arch=('x86_64' 'i686')
    url="https://example.com/my-app"
    license=('MIT')
    depends=('bash' 'python>=3.8')
    makedepends=('git')
    source=("my-app-$pkgver.tar.gz::https://example.com/releases/$pkgver/my-app.tar.gz")

    prepare() {
        cd "$srcdir/$pkgname-$pkgver"
        ./configure --prefix=/usr
    }

    build() {
        make
    }

    package() {
        make install DESTDIR="$pkgdir"
    }

    check() {
        make check
    }
    """
    local_pkgbuild_path = "/tmp/my-app/PKGBUILD"
    os.makedirs(os.path.dirname(local_pkgbuild_path), exist_ok=True)
    with open(local_pkgbuild_path, "w", encoding='utf-8') as f:
        f.write(pkgbuild_content_local)

    pkgbuild_content_upstream = """
    # Contributor: Upstream Dev
    pkgname=my-app
    pkgver=1.0.1
    pkgrel=1
    arch=('x86_64')
    url="https://example.com/my-app-new" # Changed URL
    license=('MIT')
    depends=('bash' 'python>=3.9' 'new-dep') # Added new-dep, updated python
    makedepends=('git' 'cmake') # Added cmake
    source=("my-app-$pkgver.tar.gz::https://example.com/releases/$pkgver/my-app.tar.gz")

    prepare() {
        cd "$srcdir/$pkgname-$pkgver"
        ./configure --prefix=/usr --enable-features # Added feature flag
    }

    build() {
        echo "Building new version!" # Changed build message
        make
    }

    package() {
        make install DESTDIR="$pkgdir"
        install -m644 new-file "$pkgdir/usr/share/my-app/" # Added new file
    }
    """
    upstream_pkgbuild_path = "/tmp/my-app-upstream/PKGBUILD"
    os.makedirs(os.path.dirname(upstream_pkgbuild_path), exist_ok=True)
    with open(upstream_pkgbuild_path, "w", encoding='utf-8') as f:
        f.write(pkgbuild_content_upstream)


    analyzer = PKGBUILDAnalyzer()

    print("\n--- Testing parse_pkgbuild_detailed (Local) ---")
    local_metadata = analyzer.parse_pkgbuild_detailed(local_pkgbuild_path)
    if local_metadata:
        print(f"Pkgname: {local_metadata.pkgname}")
        print(f"Pkgver: {local_metadata.pkgver}-{local_metadata.pkgrel}")
        print(f"Arch: {local_metadata.arch}")
        print(f"Depends: {local_metadata.depends}")
        print(f"Makedepends: {local_metadata.makedepends}")
        print(f"Source: {local_metadata.source}")
        print("Functions:")
        for name, func in local_metadata.functions.items():
            print(f"  - {func.name} (Lines {func.start_line}-{func.end_line}):\n{func.content[:100]}...") # Show first 100 chars
    else:
        print("Failed to parse local PKGBUILD.")


    print("\n--- Testing compare_pkgbuilds ---")
    comparison_result = analyzer.compare_pkgbuilds(local_pkgbuild_path, upstream_pkgbuild_path)
    if comparison_result:
        print(f"Has Changes: {comparison_result.has_changes}")
        print(f"Version Diff: Local={comparison_result.version_diff[0]}, Upstream={comparison_result.version_diff[1]}")
        print(f"Critical Sections Changed: {comparison_result.critical_sections_changed}")
        print("\nUnified Diff:\n")
        print(comparison_result.diff)
    else:
        print("Failed to compare PKGBUILDs.")


    print("\n--- Testing analyze_dependencies (requires pacman) ---")
    # This will assume 'bash' and 'python' are installed on the system.
    # 'new-dep' might be missing if not installed for the 'upstream' PKGBUILD.
    dep_analysis = analyzer.analyze_dependencies(upstream_pkgbuild_path)
    if dep_analysis:
        print(f"Dependencies for {dep_analysis.package_name} v{dep_analysis.current_pkgver}:")
        print(f"  Depends: {dep_analysis.dependencies.get('depends')}")
        print(f"  Makedepends: {dep_analysis.dependencies.get('makedepends')}")
        print(f"  Missing Dependencies: {dep_analysis.missing_deps}")
        print(f"  Installed Dependencies: {dep_analysis.installed_deps}")
    else:
        print("Failed to analyze dependencies.")

    print("\n--- Testing verify_version_compatibility ---")
    compatible, msg = analyzer.verify_version_compatibility(local_pkgbuild_path)
    print(f"Compatibility for local PKGBUILD: {compatible} - {msg}")

    compatible_upstream, msg_upstream = analyzer.verify_version_compatibility(upstream_pkgbuild_path)
    print(f"Compatibility for upstream PKGBUILD: {compatible_upstream} - {msg_upstream}")


    # Clean up dummy files
    os.remove(local_pkgbuild_path)
    os.rmdir(os.path.dirname(local_pkgbuild_path)) # Remove dir if empty
    os.remove(upstream_pkgbuild_path)
    os.rmdir(os.path.dirname(upstream_pkgbuild_path)) # Remove dir if empty

    print("\n--- PKGBUILDAnalyzer Test Complete ---")

File: src/gear/pkgbuild_analyzer.rs
use std::fs;
use std::path::Path;
use std::process::Command;
use std::collections::HashMap;
use regex::Regex;
use similar::{ChangeTag, TextDiff};

#[derive(Debug, Clone, Default)]
pub struct PkgbuildFunction {
    pub name: String,
    pub content: String,
    pub start_line: usize,
    pub end_line: usize,
}

#[derive(Debug, Clone, Default)]
pub struct PkgbuildMetadata {
    pub pkgname: String,
    pub pkgver: String,
    pub pkgrel: String,
    pub epoch: Option<String>,
    pub arch: Vec<String>,
    pub url: Option<String>,
    pub license: Option<String>,
    pub depends: Vec<String>,
    pub makedepends: Vec<String>,
    pub checkdepends: Vec<String>,
    pub optdepends: Vec<String>,
    pub source: Vec<String>,
    pub functions: HashMap<String, PkgbuildFunction>,
}

#[derive(Debug, Clone)]
pub struct PkgbuildComparisonResult {
    pub diff: String,
    pub has_changes: bool,
    pub critical_sections_changed: Vec<String>,
    pub version_diff: (String, String),
}

#[derive(Debug, Clone)]
pub struct DependencyAnalysisResult {
    pub package_name: String,
    pub current_pkgver: String,
    pub dependencies: HashMap<String, Vec<String>>,
    pub missing_deps: Vec<String>,
    pub installed_deps: Vec<String>,
    pub optional_deps_available: Vec<String>,
}

pub struct PKGBUILDAnalyzer {
    _array_var_re: Regex,
    _single_var_re: Regex,
    _function_re: Regex,
    _func_end_re: Regex,
}

impl PKGBUILDAnalyzer {
    pub fn new() -> Self {
        Self {
            _array_var_re: Regex::new(r"(?m)^\s*(\w+)\s*=\s*\(([^)]*)\)").unwrap(),
            _single_var_re: Regex::new(r"(?m)^\s*(\w+)\s*=\s*(.*?)\s*$").unwrap(),
            _function_re: Regex::new(r"(?m)^\s*(\w+)\s*\(\s*\)\s*\{").unwrap(),
            _func_end_re: Regex::new(r"(?m)^\s*\}").unwrap(),
        }
    }

    pub fn parse_pkgbuild_detailed(&self, pkgbuild_path: &str) -> Option<PkgbuildMetadata> {
        if !Path::new(pkgbuild_path).exists() {
            return None;
        }

        let content = fs::read_to_string(pkgbuild_path).ok()?;
        let content_no_comments = Regex::new(r"(?m)#.*$").unwrap().replace_all(&content, "").to_string();

        let mut metadata = PkgbuildMetadata {
            pkgname: "unknown".to_string(),
            pkgver: "unknown".to_string(),
            pkgrel: "1".to_string(),
            ..Default::default()
        };

        // Extract basic variables
        let var_names = vec!["pkgname", "pkgver", "pkgrel", "epoch", "url", "license", "arch"];
        for var_name in var_names {
            let pattern = format!(r"(?m)^\s*{}\s*=\s*(.*?)\s*$", var_name);
            if let Ok(re) = Regex::new(&pattern) {
                if let Some(caps) = re.captures(&content_no_comments) {
                    let value = caps[1].trim().trim_matches(|c| c == '\'' || c == '"' || c == '(' || c == ')').to_string();
                    match var_name {
                        "pkgname" => metadata.pkgname = value,
                        "pkgver" => metadata.pkgver = value,
                        "pkgrel" => metadata.pkgrel = value,
                        "epoch" => metadata.epoch = Some(value),
                        "url" => metadata.url = Some(value),
                        "license" => metadata.license = Some(value),
                        "arch" => metadata.arch = value.split_whitespace().map(|s| s.to_string()).collect(),
                        _ => {}
                    }
                }
            }
        }

        // Extract array variables
        for caps in self._array_var_re.captures_iter(&content_no_comments) {
            let var_name = &caps[1];
            let raw_content = &caps[2];
            let items: Vec<String> = raw_content.split_whitespace()
                .map(|s| s.trim_matches(|c| c == '\'' || c == '"').to_string())
                .filter(|s| !s.is_empty())
                .collect();

            match var_name {
                "depends" => metadata.depends.extend(items),
                "makedepends" => metadata.makedepends.extend(items),
                "checkdepends" => metadata.checkdepends.extend(items),
                "optdepends" => metadata.optdepends.extend(items),
                "source" => metadata.source.extend(items),
                _ => {}
            }
        }

        // Resolve variables in source
        metadata.source = metadata.source.iter()
            .map(|s| s.replace("$pkgname", &metadata.pkgname).replace("$pkgver", &metadata.pkgver))
            .collect();

        // Extract functions
        let lines: Vec<&str> = content.lines().collect();
        let mut in_function = false;
        let mut current_func_name = String::new();
        let mut func_content = String::new();
        let mut func_start_line = 0;

        for (i, line) in lines.iter().enumerate() {
            if let Some(caps) = self._function_re.captures(line) {
                if in_function {
                    metadata.functions.insert(current_func_name.clone(), PkgbuildFunction {
                        name: current_func_name.clone(),
                        content: func_content.clone(),
                        start_line: func_start_line,
                        end_line: i,
                    });
                }
                in_function = true;
                current_func_name = caps[1].to_string();
                func_content = String::new();
                func_start_line = i + 1;
            } else if in_function && self._func_end_re.is_match(line) {
                metadata.functions.insert(current_func_name.clone(), PkgbuildFunction {
                    name: current_func_name.clone(),
                    content: func_content.clone(),
                    start_line: func_start_line,
                    end_line: i + 1,
                });
                in_function = false;
                current_func_name = String::new();
                func_content = String::new();
            } else if in_function {
                func_content.push_str(line);
                func_content.push('\n');
            }
        }

        if in_function {
            metadata.functions.insert(current_func_name.clone(), PkgbuildFunction {
                name: current_func_name.clone(),
                content: func_content.clone(),
                start_line: func_start_line,
                end_line: lines.len(),
            });
        }

        Some(metadata)
    }

    pub fn compare_pkgbuilds(&self, local_path: &str, upstream_path: &str) -> Option<PkgbuildComparisonResult> {
        let local_content = fs::read_to_string(local_path).ok()?;
        let upstream_content = fs::read_to_string(upstream_path).ok()?;

        let diff = TextDiff::from_lines(&local_content, &upstream_content);
        let mut diff_str = String::new();
        for hunk in diff.unified_diff().iter_hunks() {
            for change in hunk.iter_changes() {
                match change.tag() {
                    ChangeTag::Delete => diff_str.push_str(&format!("-{}", change)),
                    ChangeTag::Insert => diff_str.push_str(&format!("+{}", change)),
                    ChangeTag::Equal => diff_str.push_str(&format!(" {}", change)),
                }
            }
        }

        let has_changes = !diff_str.is_empty();
        let mut critical_sections_changed = Vec::new();
        if has_changes {
            if diff_str.contains("source=") || diff_str.contains("prepare()") || diff_str.contains("package()") {
                critical_sections_changed.push("source".to_string());
                critical_sections_changed.push("prepare_function".to_string());
                critical_sections_changed.push("package_function".to_string());
            }
        }

        let local_meta = self.parse_pkgbuild_detailed(local_path);
        let upstream_meta = self.parse_pkgbuild_detailed(upstream_path);

        let local_ver = local_meta.map(|m| format!("{}-{}", m.pkgver, m.pkgrel)).unwrap_or_else(|| "N/A".to_string());
        let upstream_ver = upstream_meta.map(|m| format!("{}-{}", m.pkgver, m.pkgrel)).unwrap_or_else(|| "N/A".to_string());

        Some(PkgbuildComparisonResult {
            diff: diff_str,
            has_changes,
            critical_sections_changed,
            version_diff: (local_ver, upstream_ver),
        })
    }

    pub fn analyze_dependencies(&self, pkgbuild_path: &str) -> Option<DependencyAnalysisResult> {
        let metadata = self.parse_pkgbuild_detailed(pkgbuild_path)?;
        
        let mut result = DependencyAnalysisResult {
            package_name: metadata.pkgname.clone(),
            current_pkgver: metadata.pkgver.clone(),
            dependencies: HashMap::new(),
            missing_deps: Vec::new(),
            installed_deps: Vec::new(),
            optional_deps_available: Vec::new(),
        };

        result.dependencies.insert("depends".to_string(), metadata.depends.clone());
        result.dependencies.insert("makedepends".to_string(), metadata.makedepends.clone());
        result.dependencies.insert("checkdepends".to_string(), metadata.checkdepends.clone());
        result.dependencies.insert("optdepends".to_string(), metadata.optdepends.clone());

        let all_req_deps: Vec<String> = metadata.depends.iter().chain(metadata.makedepends.iter()).chain(metadata.checkdepends.iter()).cloned().collect();

        let installed_output = Command::new("pacman").args(["-Qq"]).output().ok()?;
        let installed_str = String::from_utf8_lossy(&installed_output.stdout);
        let installed_set: std::collections::HashSet<&str> = installed_str.lines().collect();

        for dep in all_req_deps {
            // Simple check (doesn't handle versioning yet)
            let dep_name = dep.split(|c| c == '>' || c == '<' || c == '=').next().unwrap_or(&dep).trim();
            if installed_set.contains(dep_name) {
                result.installed_deps.push(dep);
            } else {
                result.missing_deps.push(dep);
            }
        }

        for optdep in metadata.optdepends {
            let dep_name = optdep.split(|c| c == '>' || c == '<' || c == '=').next().unwrap_or(&optdep).trim();
            if installed_set.contains(dep_name) {
                result.optional_deps_available.push(optdep);
            }
        }

        Some(result)
    }

    pub fn verify_version_compatibility(&self, pkgbuild_path: &str) -> (bool, String) {
        let metadata = match self.parse_pkgbuild_detailed(pkgbuild_path) {
            Some(m) => m,
            None => return (false, "Could not parse PKGBUILD metadata.".to_string()),
        };

        let arch_output = Command::new("uname").arg("-m").output();
        let current_arch = arch_output.map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string()).unwrap_or_else(|_| "unknown".to_string());

        if !metadata.arch.contains(&"any".to_string()) && !metadata.arch.contains(&current_arch) {
            return (false, format!("Architecture '{}' not supported (supports: {:?})", current_arch, metadata.arch));
        }

        (true, "Version and architecture appear compatible.".to_string())
    }
}
